### PR TITLE
Tally trigger

### PR DESCRIPTION
**Pull request by shenqicang**
This is the OpenMC code with tally precision triggers and corresponding documentation. The trigger can be set for combined k_effective and tallies. The types of trigger can be set to relative error, standard deviation and variance. In some situation, it can predict the number of batches to satisfy the trigger. As far as  I know, there's no problem. Hope everyone can give advice and guidance. This Pull request fulfills issue #264.

### CODE REVIEW

**Comment by wbinventor** [comment]
Quick update on this PR - I synchronized this to the latest version of the "develop" branch, and @shenqicang has done some work to reduce the memory requirements for the tally trigger convergence check. I plan to completely run through this PR again soon in the hopes of finishing the merge before @shenqicang presents it the ANS summer meeting. @paulromano - are you still keen on merging this into develop?

File: src/input_xml.F90 [hunk 1]
```diff
@@ -2818,6 +2935,207 @@ subroutine read_tallies_xml()
              &// trim(to_str(t % id)) // ".")
       end if
 
+    !Read the trigger information
```
Comment (smharper) [inline src/input_xml.F90]: The code here is missing two spaces of indentation.

File: src/input_xml.F90 [hunk 2]
```diff
@@ -2818,6 +2935,207 @@ subroutine read_tallies_xml()
...
+            case ('all')
+              if (t % n_user_triggers /= 1) then
+                call fatal_error("Cannot set trigger for all and other scoring &
+                       & functions in the same tally. Separate scoring &
+                       & functions into distinct tallies")
+              else
+                t % trigger_for_all = .true.
+              end if
+            case ('flux')
+              t % score(trig_ind) % position =  find_score % get_key(SCORE_FLUX)
```
Comment (smharper) [inline src/input_xml.F90]: Some of these `get_key` lines need to be split so they fit within 80 chars

File: src/input_xml.F90 [hunk 3]
```diff
@@ -2818,6 +2935,207 @@ subroutine read_tallies_xml()
...
+                    n_order = MAX_ANG_ORDER
+                  end if
+                  score_name = trim(MOMENT_N_STRS(imomstr)) // "n"
+                  exit
+                end if
+              end do
+            end if
+            t % score(trig_ind) % score_name = score_name
+
+            select case (trim(score_name))
```
Comment (smharper) [inline src/input_xml.F90]: Do we need this `select case` block at all?  It looks like we may be able to replace it with something like `t % score(trig_ind) % position = find_score % get_key(t % score_bins(j))`

**Comment by smharper** [comment]
Has anything been done to address fixed source calculations?

**Comment by paulromano** [comment]
@wbinventor I would still like to get this merged, but the implementation needs to be complete and sound. I'll take a look and see if the memory usage changes are saitsfactory. As @smharper alluded to, we need support for fixed source calculations as well. Also, the history on this pull request is a mess with lots of files being added and deleted and nonsensical commit messages, so I will probably request that once it's in a state where it's ready to be merged in, we squash it into a single commit on a new branch and then merge that instead.

**Comment by wbinventor** [comment]
@shenqicang - I will review your code this weekend and comment on anything that I see that needs to be changed. Although I am not familiar with how the fixed source calculation is setup, I do not see any reason for it to be that difficult to extend the tally triggers to that problem domain. Of course as @paulromano mentioned we will need to squash most if not all of your commits down to a single commit since most of the log messages do not make any sense.

**Comment by wbinventor** [comment]
Make sure to merge your code with the latest version of the develop branch. We also need to extend the new Python API to include tally trigger support, which I may do and then submit a PR to your fork.

**Comment by wbinventor** [comment]
This has been replaced by PR #381.

**Pull request closed without merging**